pkg/act/act.go

```diff
@@ -0,0 +1,303 @@
+		}
+		svc = ref.String()
+		return
+	}
+	return &ACT{
+		log:          log,
+		src:          src,
+		metrics:      metrics,
+		keyToStrings: kts,
+		mux:          new(lock.Mutex),
+		tracker:      tracker,
+	}
+}
+
+// callback processes an ACT map entry.
+//
+// It will create the new metrics series if needed. In this case metrics won't
+// be presented as the number of new connections can't be calculated yet.
+func (a *ACT) callback(key *act.ActiveConnectionTrackerKey, value *act.ActiveConnectionTrackerValue) {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+
+	entry, ok := a.tracker[key.Zone][key.SvcID]
+	if !ok {
+		zone, svc, err := a.keyToStrings(key)
+		if err != nil {
+			a.log.Debug("Failed to construct metrics map key in callback", "from-key", key.String())
+			return
+		}
+		a.tracker[key.Zone][key.SvcID] = &actMetric{
+			opened:      uint64(value.Opened),
+			closed:      uint64(value.Closed),
+			labelValues: []string{zone, svc},
+		}
+		return
+	}
+
+	opened, closed := uint64(value.Opened), uint64(value.Closed)
+	if opened == entry.opened && closed == entry.closed && entry.newFailed == 0 {
+		a.metrics.New.WithLabelValues(entry.labelValues...).Set(0)
+		a.metrics.Failed.WithLabelValues(entry.labelValues...).Set(0)
+		return
+	}
+	scopedLog := a.log.With("svc", key.SvcID, "zone", key.Zone)
+
+	// Opened/Closed are 32-bit values, so they can roll-over. Adjust
+	if opened < entry.opened {
+		opened += math.MaxUint32
+	}
+	if closed < entry.closed {
+		closed += math.MaxUint32
+	}
+
+	entry.failed += entry.newFailed
+	a.metrics.Failed.WithLabelValues(entry.labelValues...).Set(float64(entry.newFailed))
+
+	active := opened - (closed + entry.failed)
+	if sumClosed := (closed + entry.failed); sumClosed > opened {
+		scopedLog.Error("Unexpected closed+failed", "got", sumClosed, "want", opened)
+		opened = closed + entry.failed
+	}
+	a.metrics.Active.WithLabelValues(entry.labelValues...).Set(float64(active))
+
+	new := opened - entry.opened
+	a.metrics.New.WithLabelValues(entry.labelValues...).Set(float64(new))
+
+	entry.opened = opened
+	entry.closed = closed
+	entry.newFailed = 0
+	entry.updated = time.Now()
+}
+
+func (a *ACT) update(ctx context.Context) error {
+	start := time.Now()
+	err := a.src.IterateWithCallback(ctx, a.callback)
+	if err != nil {
+		return fmt.Errorf("iterate over %q: %w", act.ActiveConnectionTrackingMapName, err)
+	}
+	a.metrics.ProcessingTime.Observe(time.Since(start).Seconds())
+	return nil
+}
+
+func (a *ACT) dropEntry(zone uint8, svc uint16) {
+	entry := a.tracker[zone][svc]
+	a.metrics.New.DeleteLabelValues(entry.labelValues...)
+	a.metrics.Active.DeleteLabelValues(entry.labelValues...)
+	a.metrics.Failed.DeleteLabelValues(entry.labelValues...)
+
+	mapKey := &act.ActiveConnectionTrackerKey{SvcID: svc, Zone: zone}
+	err := a.src.Delete(mapKey)
+	if err != nil {
+		a.log.Debug("Failed to delete ACT map entry", "key", mapKey.String())
+	}
+	delete(a.tracker[zone], svc)
```

Comment:
Should we delete the userspace entry if the bpf map key failed to be deleted?

Reply:
I don't know what is a typical error for deletion (we already ignore not-exists errors). If they are transient, then I think we should keep the userspace entry in order to retry on the next cleanup.

In the current scenario, the userspace entry will be recreated (without metrics) and then cleaned up again (if stale)

Reply:
I have yet another PR prepared in which I will move BPF map entry deletion to "service reconciler" so this will be non-issue